data-plane/control-protocol/src/main/java/dev/knative/eventing/control/protocol/impl/TCPControlServerVerticle.java

```diff
@@ -0,0 +1,183 @@
+      .setPort(port)
+      .setTcpKeepAlive(true);
+  }
+
+  private void newConnectionHandler(NetSocket netSocket) {
+    if (actualConnection != null) {
+      // Close the previous connection,
+      // the control protocol assumes only one connection is open at any given time
+      actualConnection.close();
+    }
+    actualConnection = netSocket;
+    actualConnection.handler(new ControlMessageParser(this::handleInboundMessage));
+    actualConnection.exceptionHandler(t -> logger.error("Error in connection", t));
+
+    // Dequeue the elements waiting for a connection
+    while (!enqueuedWaitingForConnection.isEmpty()) {
+      writeOnConnection(enqueuedWaitingForConnection.poll());
+    }
+  }
+
+  private void handleInboundMessage(ControlMessage message) {
+    if (message.opCode() == ControlMessage.ACK_OP_CODE) {
+      handleInboundAck(message);
+    } else {
+      emitMessageOnEventBus(message);
+    }
+  }
+
+  private void handleInboundAck(ControlMessage ackMessage) {
+    Message<ControlMessage> message = this.toAck.get(ackMessage.uuid());
+    if (message == null) {
+      logger.warn("Received an ack for an unknown uuid {}", ackMessage.uuid());
+    } else {
+      message.reply(ackMessage);
+    }
+  }
+
+  private void emitMessageOnEventBus(ControlMessage message) {
+    this.serializedEventBusRequester.request(message)
+      .onFailure(t -> logger
+        .error("Cannot route the incoming control message to {}: {}", this.incomingMessageAddress, message, t))
+      .onSuccess(m -> sendAckBack(message.uuid(), m.body()));
+  }
+
+  private void sendAckBack(UUID uuid, Object body) {
+    ControlMessageImpl.Builder builder = new ControlMessageImpl.Builder()
+      .setOpCode(ControlMessage.ACK_OP_CODE)
+      .setUuid(uuid);
+
+    if (body != null) {
+      if (body instanceof Throwable) {
+        builder
+          .setLengthAndPayload(
+            Buffer.buffer(((Throwable) body).getMessage())
+          );
+      } else {
+        builder
+          .setLengthAndPayload(Buffer.buffer(String.valueOf(body)));
+      }
+    }
+
+    ControlMessage message = builder.build();
+    writeOnConnection(message);
+  }
+
+  private void handleOutboundMessage(Message<ControlMessage> message) {
+    ControlMessage controlMessage = message.body();
+    writeOnConnection(controlMessage);
+    this.toAck.put(controlMessage.uuid(), message);
+  }
+
+  private void writeOnConnection(ControlMessage message) {
+    if (actualConnection == null) {
+      // Enqueue the message
+      this.enqueuedWaitingForConnection.offer(message);
+      return;
+    }
+    this.internalWriteOnConnection(message, message.toBuffer());
+  }
+
+  private void internalWriteOnConnection(ControlMessage message, Buffer buffer) {
+    actualConnection.write(buffer)
+      .onSuccess(v -> logger.debug("Successfully sent {}", message))
+      .onFailure(t -> {
+        logger.error("Cannot dispatch message {}", message, t);
+        // TODO should we retry here?
```

Comment:
I think we should have "a business level" retry policy, maybe some messages are important others might be less

Reply:
and at this layer, we don't have enough context. :thinking:

Reply:
I'm specifically thinking at stuff like:

"I cannot run a producer because xyz" vs "I cannot apply a retry policy update"

Reply:
Yes, sounds good to me let's thing to this problem later, I was just sharing some random thoughts.